app/src/battery.c

```diff
@@ -36,17 +40,23 @@ static int zmk_battery_update(const struct device *battery) {
         return rc;
     }
 
-    LOG_DBG("Setting BAS GATT battery level to %d.", state_of_charge.val1);
+    if (last_state_of_charge != state_of_charge.val1) {
+        LOG_DBG("Setting BAS GATT battery level to %d.", state_of_charge.val1);
 
-    rc = bt_bas_set_battery_level(state_of_charge.val1);
+        rc = bt_bas_set_battery_level(state_of_charge.val1);
 
-    if (rc != 0) {
-        LOG_WRN("Failed to set BAS GATT battery level (err %d)", rc);
-        return rc;
+        if (rc != 0) {
+            LOG_WRN("Failed to set BAS GATT battery level (err %d)", rc);
+            return rc;
+        }
+
+        rc = ZMK_EVENT_RAISE(new_zmk_battery_state_changed(
+            (struct zmk_battery_state_changed){.state_of_charge = state_of_charge.val1}));
+
+        last_state_of_charge = state_of_charge.val1;
```

Comment:
We should set the value *before* raising the event, in case any listeners ignore the event payload and instead call the newly added API to retrieve the value.

Reply:
done